### PR TITLE
sqlbase: multirowfetcher stops decoding a row when it doesn't need more data

### LINKED ISSUE #21443: sqlbase: teach multirowfetcher to stop decoding a row when it doesn't need more data
Currently, MultiRowFetcher reads through the entire value of a kv if it needs *any* of the columns from that value. If a column is not necessary, it just reads the tag of that column to figure out how to skip ahead in the data.

However, reading that tag has a cost. For rows with many columns, a query that only needs the first value in the row will still cause the fetcher to read the lengths of *all* of the columns in the row.

We should keep track of when we've got all the data that we need and quit early at that point.

I noticed this while profiling TPCC's stock level transaction, which emits the following query:

```
				SELECT COUNT(DISTINCT(s_i_id))
				FROM order_line
				JOIN stock
				ON s_i_id=ol_i_id
				WHERE ol_w_id = $1
				  AND ol_d_id = $2
				  AND ol_o_id BETWEEN $3 - 20 AND $3 - 1
				  AND s_w_id = $1
				  AND s_quantity < $4

```

`stock` is a table with 17 columns. This query only touches the first one that isn't in the key: `s_quantity`. Judging by pprof output, MRF nevertheless spends at least 5% of its time reading the lengths of the rest of those columns to skip over them in the value portion of the kv.

### PR DESCRIPTION
**Pull request by jordanlewis**
Previously, if the row fetcher needed any of the columns from the value
portion of the keys in a row, it would iterate over the whole value
column by column, decoding the needed columns and skipping unneeded
columns by reading their length. This would occur even after all of the
needed columns were decoded.

This is inefficient for certain kinds of tables and queries. For
example, consider a very wide table and a query that only needs values
from the beginning of the table. In that case, the row fetcher would
read the length of every column value even once it had gotten all of the
column values that it needed.

After this change, the row fetcher keeps track of the number of column
values it needs, and quits decoding early once it has read all of the
required columns. The same is done in `finalizeRow`, which now quits
early when all needed columns are set.

```
name                                old time/op    new time/op    delta
WideTableIgnoreColumns/Cockroach-8    21.5ms ± 2%    14.8ms ± 2%  -31.09%  (p=0.000 n=10+10)

name                                old alloc/op   new alloc/op   delta
WideTableIgnoreColumns/Cockroach-8    2.14MB ± 1%    2.12MB ± 0%   -1.02%  (p=0.000 n=10+9)

name                                old allocs/op  new allocs/op  delta
WideTableIgnoreColumns/Cockroach-8     1.41k ±54%     1.33k ±29%     ~     (p=0.239 n=10+10)
```

Release note (performance improvement): Queries that only need to read
part of a table with many columns are more efficient.

Fixes #21443.

### CODE REVIEW

**Comment by jordanlewis** [comment]
Added another commit that does a similar optimization in `finalizeRow`, which previously looped over all rows as well. It gave another 5% improvement. I updated the performance numbers in the PR message to include both commits.

**Review by RaduBerinde — APPROVED** [review]
LGTM

In the last commit message, I think you mean columns ("if we already know we've gotten all of the rows that we need").

**Comment by jordanlewis** [comment]
TFTR, yeah I said rows instead of columns twice here. Thanks!

File: pkg/sql/sqlbase/multirowfetcher.go [hunk 1]
```diff
@@ -1093,6 +1097,7 @@ func (mrf *MultiRowFetcher) finalizeRow() error {
 			table.row[i] = EncDatum{
 				Datum: tree.DNull,
 			}
+			mrf.valueColsFound++
```
[the remark this answers is not in the document]
Comment (jordanlewis) [reply pkg/sql/sqlbase/multirowfetcher.go]: Done.

**Pull request merged**